base/linalg/dense.jl

```diff
@@ -29,6 +29,21 @@ function scale!(X::Array{T}, s::Real) where T<:BlasComplex
     X
 end
 
+function isone(x::StridedMatrix)
+    m, n = size(x)
+    m != n && return false # only square matrices can satisfy x == one(x)
+    for i in 1:m
+        for j in i:m
+            if i == j
+                @inbounds !isone(x[i,i]) && return false
+            else
+                @inbounds (!iszero(x[i,j]) || !iszero(x[j,i])) && return false
```

Comment:
You're right. I tried:
```julia
function isone_friendly(x::StridedMatrix)
    m, n = size(x)
    m != n && return false # only square matrices can satisfy x == one(x)
    @inbounds for i in 1:m, j in 1:m
        if i == j
            isone(x[i,i]) || return false
        else
            iszero(x[j,i]) || return false
        end
    end
    return true
end

function isone_unfriendly(x::StridedMatrix)
... [j, i] -> [i, j]
end

function isone_triangular(x::StridedMatrix)
... for j in i:m
end
```
with the benchmark results:
```julia
julia> const C = eye(4000, 4000)
julia> @benchmark isone_friendly(C)
...
  median time:      14.152 ms (0.00% GC)
...

julia> @benchmark isone_unfriendly(C)
...
  median time:      52.547 ms (0.00% GC)
...

julia> @benchmark isone_triangular(C)
...
  median time:      24.796 ms (0.00% GC)
...
```
But for `C = eye(400, 400)`, the triangular one is still faster. Should there be some size check to see if it fits in the cache, and if so branch into one method or the other?